File: api_server/db_repository.py
"""WMS data layer — SQL Server backend (Rohteillager database)."""
from datetime import date, datetime
from sql_client import get_cursor


def _strip(val):
    return val.strip() if isinstance(val, str) else val


def _row_to_booking(row, columns):
    raw = dict(zip(columns, row))
    return {
        "id": str(raw["Rownumber"]),
        "location_code": _strip(raw.get("Lagerort", "")),
        "item_number": _strip(raw.get("Artikelnummer", "")),
        "item_type": _strip(raw.get("Art", "")),
        "batch_number": _strip(raw.get("Lieferschein", "")),
        "qty": int(_strip(raw.get("Menge", "0")) or 0),
        "description": _strip(raw.get("Bezeichnung", "")),
        "booked_by": _strip(raw.get("Zusatz1", "")),
        "booked_at": raw["Datum"].isoformat() if isinstance(raw.get("Datum"), (date, datetime)) else str(raw.get("Datum", "")),
        "status": "Active",
        "batch_charge": _strip(raw.get("Chargennummer", "")),
    }


def _row_to_movement(row, columns):
    raw = dict(zip(columns, row))
    return {
        "id": str(raw["Rownumber"]),
        "action": _strip(raw.get("Bemerkung", "")),
        "location_code": _strip(raw.get("Lagerort", "")),
        "item_number": _strip(raw.get("Artikelnummer", "")),
        "batch_number": _strip(raw.get("Lieferschein", "")),
        "qty": int(_strip(raw.get("Menge", "0")) or 0),
        "user": _strip(raw.get("Zusatz1", "")),
        "timestamp": raw["Datum"].isoformat() if isinstance(raw.get("Datum"), (date, datetime)) else str(raw.get("Datum", "")),
        "description": _strip(raw.get("Bezeichnung", "")),
    }


def _columns(cursor):
    return [desc[0] for desc in cursor.description]


# ── Users ──

def init_db():
    with get_cursor() as cur:
        cur.execute("""
            IF NOT EXISTS (SELECT * FROM sys.tables WHERE name = 'users')
            CREATE TABLE users (
                username NCHAR(20) NOT NULL PRIMARY KEY,
                password_hash NVARCHAR(200) NOT NULL,
                role NCHAR(20) NOT NULL
            )
        """)


def get_user(username):
    with get_cursor() as cur:
        cur.execute("SELECT username, password_hash, role FROM users WHERE RTRIM(username) = ?", username)
        row = cur.fetchone()
        if row is None:
            return None
        return {
            "id": _strip(row[0]),
            "username": _strip(row[0]),
            "password_hash": _strip(row[1]),
            "role": _strip(row[2]),
        }


def get_all_users():
    with get_cursor() as cur:
        cur.execute("SELECT username, role FROM users ORDER BY username")
        return [
            {"id": _strip(r[0]), "username": _strip(r[0]), "role": _strip(r[1])}
            for r in cur.fetchall()
        ]


def create_user(username, password_hash, role):
    with get_cursor() as cur:
        cur.execute("INSERT INTO users (username, password_hash, role) VALUES (?, ?, ?)",
                     username, password_hash, role)


def update_user(user_id, username=None, password_hash=None, role=None):
    parts, params = [], []
    if username:
        parts.append("username = ?")
        params.append(username)
    if password_hash:
        parts.append("password_hash = ?")
        params.append(password_hash)
    if role:
        parts.append("role = ?")
        params.append(role)
    if not parts:
        return
    params.append(user_id)
    with get_cursor() as cur:
        cur.execute(f"UPDATE users SET {', '.join(parts)} WHERE RTRIM(username) = ?", *params)


def delete_user(user_id):
    with get_cursor() as cur:
        cur.execute("DELETE FROM users WHERE RTRIM(username) = ?", user_id)


# ── Inventory (rohteillager) ──

def get_cell_contents(location_code):
    with get_cursor() as cur:
        cur.execute(
            "SELECT * FROM rohteillager WHERE RTRIM(Lagerort) = ? AND CAST(RTRIM(Menge) AS INT) > 0",
            location_code,
        )
        cols = _columns(cur)
        return sorted(
            [_row_to_booking(r, cols) for r in cur.fetchall()],
            key=lambda x: x.get("booked_at", ""),
        )


def get_active_slot_count(location_code):
    with get_cursor() as cur:
        cur.execute(
            "SELECT COUNT(*) FROM rohteillager WHERE RTRIM(Lagerort) = ? AND CAST(RTRIM(Menge) AS INT) > 0",
            location_code,
        )
        return cur.fetchone()[0]


def book_in(location_code, item_number, item_type, batch_number, qty, description, user):
    with get_cursor() as cur:
        cur.execute(
            """INSERT INTO rohteillager
               (Artikelnummer, Bezeichnung, Lagerort, Menge, Datum, Art, Lieferschein, Zusatz1)
               VALUES (?, ?, ?, ?, CAST(GETDATE() AS DATE), ?, ?, ?)""",
            item_number[:20], description[:40], location_code[:20], str(qty)[:10], item_type[:20], batch_number[:20], user[:20],
        )
        cur.execute("SELECT SCOPE_IDENTITY()")
        new_id = cur.fetchone()[0]
        booking_id = str(int(new_id)) if new_id else "0"

        cur.execute(
            """INSERT INTO journal
               (Artikelnummer, Bezeichnung, Lagerort, Menge, Datum, Art, Lieferschein, Zusatz1, Bemerkung)
               VALUES (?, ?, ?, ?, CAST(GETDATE() AS DATE), ?, ?, ?, ?)""",
            item_number[:20], description[:40], location_code[:20], str(qty)[:10], item_type[:20], batch_number[:20], user[:20],
            f"Ein {qty} Stk"[:20],
        )
        return booking_id


def consume(booking_id, qty, user):
    with get_cursor() as cur:
        cur.execute("SELECT * FROM rohteillager WHERE Rownumber = ?", int(booking_id))
        cols = _columns(cur)
        row = cur.fetchone()
        if row is None:
            return False
        booking = _row_to_booking(row, cols)
        current_qty = booking["qty"]
        new_qty = current_qty - qty
        consumed_qty = qty if new_qty > 0 else current_qty

        if new_qty <= 0:
            cur.execute("DELETE FROM rohteillager WHERE Rownumber = ?", int(booking_id))
        else:
            cur.execute(
                "UPDATE rohteillager SET Menge = ? WHERE Rownumber = ?",
                str(new_qty), int(booking_id),
            )

        cur.execute(
            """INSERT INTO journal
               (Artikelnummer, Bezeichnung, Lagerort, Menge, Datum, Art, Lieferschein, Zusatz1, Bemerkung)
               VALUES (?, ?, ?, ?, CAST(GETDATE() AS DATE), ?, ?, ?, ?)""",
            booking["item_number"][:20], booking["description"][:40], booking["location_code"][:20],
            str(consumed_qty)[:10], booking["item_type"][:20], booking["batch_number"][:20], user[:20],
            f"Aus -{consumed_qty} Stk"[:20],
        )
        return True


# ── Zone / shelf queries ──

def get_zone_occupancy(zone_code, shelf_codes):
    if not shelf_codes:
        return 0
    placeholders = " OR ".join(["RTRIM(Lagerort) LIKE ?"] * len(shelf_codes))
    params = [f"{sc}-%" for sc in shelf_codes]
    with get_cursor() as cur:
        cur.execute(
            f"""SELECT COUNT(DISTINCT RTRIM(Lagerort))
                FROM rohteillager
                WHERE CAST(RTRIM(Menge) AS INT) > 0 AND ({placeholders})""",
            *params,
        )
        return cur.fetchone()[0]


def get_occupied_cells(shelf_code):
    with get_cursor() as cur:
        cur.execute(
            """SELECT DISTINCT RTRIM(Lagerort)
               FROM rohteillager
               WHERE RTRIM(Lagerort) LIKE ? AND CAST(RTRIM(Menge) AS INT) > 0""",
            f"{shelf_code}-%",
        )
        return [row[0] for row in cur.fetchall()]


# ── Audit / journal ──

def get_recent_movements(limit=20):
    with get_cursor() as cur:
        cur.execute(f"SELECT TOP {int(limit)} * FROM journal ORDER BY Rownumber DESC")
        cols = _columns(cur)
        return [_row_to_movement(r, cols) for r in cur.fetchall()]


# ── Search ──

def search_bookings(query):
    pattern = f"%{query}%"
    with get_cursor() as cur:
        cur.execute(
            """SELECT * FROM rohteillager
               WHERE CAST(RTRIM(Menge) AS INT) > 0
                 AND (RTRIM(Artikelnummer) LIKE ?
                   OR RTRIM(Bezeichnung) LIKE ?
                   OR RTRIM(Lagerort) LIKE ?
                   OR RTRIM(Lieferschein) LIKE ?
                   OR RTRIM(Chargennummer) LIKE ?)
               ORDER BY Datum DESC""",
            pattern, pattern, pattern, pattern, pattern,
        )
        cols = _columns(cur)
        return [_row_to_booking(r, cols) for r in cur.fetchall()]

File: tablet_android/app/src/main/java/com/wms/gridlocator/data/DbRepository.kt
package com.wms.gridlocator.data

import android.util.Log
import java.sql.ResultSet
import java.sql.Statement

object DbRepository {

    fun initDb() {
        SqlClient.withWms { conn ->
            Log.d("DbRepository", "Initializing DB...")

            conn.createStatement().use { stmt ->
                stmt.executeUpdate(
                    """IF NOT EXISTS (SELECT * FROM sys.tables WHERE name = 'users' AND schema_id = SCHEMA_ID('dbo'))
                       CREATE TABLE dbo.users (
                           username NCHAR(20) NOT NULL PRIMARY KEY,
                           password_hash NVARCHAR(200) NOT NULL,
                           role NCHAR(20) NOT NULL
                       )"""
                )
                Log.d("DbRepository", "dbo.users table ensured.")
            }

            val users = listOf(
                Triple("admin", "admin123", "Admin"),
                Triple("logistics1", "log123", "Logistics"),
                Triple("logistics2", "log123", "Logistics"),
                Triple("consumer1", "con123", "Consumer")
            )

            users.forEach { (user, pass, role) ->
                conn.prepareStatement("SELECT COUNT(*) FROM dbo.users WHERE RTRIM(username) = ?").use { ps ->
                    ps.setString(1, user)
                    ps.executeQuery().use { rs ->
                        if (rs.next() && rs.getInt(1) == 0) {
                            Log.d("DbRepository", "Seeding user: $user")
                            val hash = org.mindrot.jbcrypt.BCrypt.hashpw(pass, org.mindrot.jbcrypt.BCrypt.gensalt())
                            conn.prepareStatement("INSERT INTO dbo.users (username, password_hash, role) VALUES (?, ?, ?)").use { insert ->
                                insert.setString(1, user)
                                insert.setString(2, hash)
                                insert.setString(3, role)
                                insert.executeUpdate()
                            }
                        }
                    }
                }
            }

            Log.d("DbRepository", "DB initialization complete.")
        }
    }

    fun getUser(username: String): Map<String, String>? {
        return SqlClient.withWms { conn ->
            conn.prepareStatement(
                "SELECT username, password_hash, role FROM dbo.users WHERE RTRIM(username) = ?"
            ).use { ps ->
                ps.setString(1, username)
                ps.executeQuery().use { rs ->
                    if (rs.next()) {
                        val foundUser = rs.getString("username").trim()
                        Log.d("DbRepository", "User found: $foundUser")
                        mapOf(
                            "id" to foundUser,
                            "username" to foundUser,
                            "password_hash" to rs.getString("password_hash").trim(),
                            "role" to rs.getString("role").trim()
                        )
                    } else {
                        Log.w("DbRepository", "User not found: $username")
                        null
                    }
                }
            }
        }
    }

    // ── Cell contents ──

    fun getCellContents(locationCode: String): List<Booking> {
        return SqlClient.withWms { conn ->
            conn.prepareStatement(
                "SELECT * FROM dbo.rohteillager WHERE RTRIM(Lagerort) = ? AND CAST(RTRIM(Menge) AS INT) > 0"
            ).use { ps ->
                ps.setString(1, locationCode)
                ps.executeQuery().use { rs ->
                    val bookings = mutableListOf<Booking>()
                    while (rs.next()) bookings.add(rowToBooking(rs))
                    bookings.sortedBy { it.bookedAt }
                }
            }
        }
    }

    fun getActiveSlotCount(locationCode: String): Int {
        return SqlClient.withWms { conn ->
            conn.prepareStatement(
                "SELECT COUNT(*) FROM dbo.rohteillager WHERE RTRIM(Lagerort) = ? AND CAST(RTRIM(Menge) AS INT) > 0"
            ).use { ps ->
                ps.setString(1, locationCode)
                ps.executeQuery().use { rs ->
                    if (rs.next()) rs.getInt(1) else 0
                }
            }
        }
    }

    // ── Book-in ──

    fun bookIn(
        locationCode: String, itemNumber: String, itemType: String,
        batchNumber: String, qty: Int, description: String, user: String
    ): String {
        return SqlClient.withWms { conn ->
            val bookingId: String

            conn.prepareStatement(
                """INSERT INTO dbo.rohteillager
                   (Artikelnummer, Bezeichnung, Lagerort, Menge, Datum, Art, Lieferschein, Zusatz1)
                   VALUES (?, ?, ?, ?, CAST(GETDATE() AS DATE), ?, ?, ?)""",
                Statement.RETURN_GENERATED_KEYS
            ).use { ps ->
                ps.setString(1, itemNumber.take(20))
                ps.setString(2, description.take(40))
                ps.setString(3, locationCode.take(20))
                ps.setString(4, qty.toString().take(10))
                ps.setString(5, itemType.take(20))
                ps.setString(6, batchNumber.take(20))
                ps.setString(7, user.take(20))
                ps.executeUpdate()

                bookingId = ps.generatedKeys.use { keys ->
                    if (keys.next()) keys.getInt(1).toString() else "0"
                }
            }

            conn.prepareStatement(
                """INSERT INTO dbo.journal
                   (Artikelnummer, Bezeichnung, Lagerort, Menge, Datum, Art, Lieferschein, Zusatz1, Bemerkung)
                   VALUES (?, ?, ?, ?, CAST(GETDATE() AS DATE), ?, ?, ?, ?)"""
            ).use { ps ->
                ps.setString(1, itemNumber.take(20))
                ps.setString(2, description.take(40))
                ps.setString(3, locationCode.take(20))
                ps.setString(4, qty.toString().take(10))
                ps.setString(5, itemType.take(20))
                ps.setString(6, batchNumber.take(20))
                ps.setString(7, user.take(20))
                ps.setString(8, "Ein $qty Stk".take(20))
                ps.executeUpdate()
            }

            bookingId
        }
    }

    // ── Consume ──

    fun consume(bookingId: String, qty: Int, user: String): Boolean {
        return SqlClient.withWms { conn ->
            val booking = conn.prepareStatement(
                "SELECT * FROM dbo.rohteillager WHERE Rownumber = ?"
            ).use { ps ->
                ps.setInt(1, bookingId.toInt())
                ps.executeQuery().use { rs ->
                    if (rs.next()) rowToBooking(rs) else return@withWms false
                }
            }

            val currentQty = booking.qty
            val newQty = currentQty - qty
            val consumedQty = if (newQty > 0) qty else currentQty

            if (newQty <= 0) {
                conn.prepareStatement("DELETE FROM dbo.rohteillager WHERE Rownumber = ?").use { ps ->
                    ps.setInt(1, bookingId.toInt())
                    ps.executeUpdate()
                }
            } else {
                conn.prepareStatement("UPDATE dbo.rohteillager SET Menge = ? WHERE Rownumber = ?").use { ps ->
                    ps.setString(1, newQty.toString())
                    ps.setInt(2, bookingId.toInt())
                    ps.executeUpdate()
                }
            }

            conn.prepareStatement(
                """INSERT INTO dbo.journal
                   (Artikelnummer, Bezeichnung, Lagerort, Menge, Datum, Art, Lieferschein, Zusatz1, Bemerkung)
                   VALUES (?, ?, ?, ?, CAST(GETDATE() AS DATE), ?, ?, ?, ?)"""
            ).use { ps ->
                ps.setString(1, booking.itemNumber.take(20))
                ps.setString(2, booking.description.take(40))
                ps.setString(3, booking.locationCode.take(20))
                ps.setString(4, consumedQty.toString().take(10))
                ps.setString(5, booking.itemType.take(20))
                ps.setString(6, booking.batchNumber.take(20))
                ps.setString(7, user.take(20))
                ps.setString(8, "Aus -$consumedQty Stk".take(20))
                ps.executeUpdate()
            }

            true
        }
    }

    // ── Relocation ──

    fun searchStock(itemNumber: String): List<Booking> {
        return SqlClient.withWms { conn ->
            conn.prepareStatement(
                """SELECT * FROM dbo.rohteillager
                   WHERE RTRIM(Artikelnummer) LIKE ? AND CAST(RTRIM(Menge) AS INT) > 0
                   ORDER BY Lagerort ASC, Datum ASC, Rownumber ASC"""
            ).use { ps ->
                ps.setString(1, "%$itemNumber%")
                ps.executeQuery().use { rs ->
                    val results = mutableListOf<Booking>()
                    while (rs.next()) results.add(rowToBooking(rs))
                    results
                }
            }
        }
    }

    fun relocate(sourceBookingId: Int, destLocationCode: String, qty: Int, user: String): String {
        return SqlClient.withWms { conn ->
            val source = conn.prepareStatement(
                "SELECT * FROM dbo.rohteillager WHERE Rownumber = ?"
            ).use { ps ->
                ps.setInt(1, sourceBookingId)
                ps.executeQuery().use { rs ->
                    if (rs.next()) rowToBooking(rs) else throw IllegalStateException("Source booking not found")
                }
            }

            val actualQty = minOf(qty, source.qty)
            val newSourceQty = source.qty - actualQty

            if (newSourceQty <= 0) {
                conn.prepareStatement("DELETE FROM dbo.rohteillager WHERE Rownumber = ?").use { ps ->
                    ps.setInt(1, sourceBookingId)
                    ps.executeUpdate()
                }
            } else {
                conn.prepareStatement("UPDATE dbo.rohteillager SET Menge = ? WHERE Rownumber = ?").use { ps ->
                    ps.setString(1, newSourceQty.toString())
                    ps.setInt(2, sourceBookingId)
                    ps.executeUpdate()
                }
            }

            val bookingId: String
            conn.prepareStatement(
                """INSERT INTO dbo.rohteillager
                   (Artikelnummer, Bezeichnung, Lagerort, Menge, Datum, Art, Lieferschein, Zusatz1)
                   VALUES (?, ?, ?, ?, CAST(GETDATE() AS DATE), ?, ?, ?)""",
                Statement.RETURN_GENERATED_KEYS
            ).use { ps ->
                ps.setString(1, source.itemNumber.take(20))
                ps.setString(2, source.description.take(40))
                ps.setString(3, destLocationCode.take(20))
                ps.setString(4, actualQty.toString().take(10))
                ps.setString(5, source.itemType.take(20))
                ps.setString(6, source.batchNumber.take(20))
                ps.setString(7, user.take(20))
                ps.executeUpdate()
                bookingId = ps.generatedKeys.use { keys ->
                    if (keys.next()) keys.getInt(1).toString() else "0"
                }
            }

            conn.prepareStatement(
                """INSERT INTO dbo.journal
                   (Artikelnummer, Bezeichnung, Lagerort, Menge, Datum, Art, Lieferschein, Zusatz1, Bemerkung)
                   VALUES (?, ?, ?, ?, CAST(GETDATE() AS DATE), ?, ?, ?, ?)"""
            ).use { ps ->
                ps.setString(1, source.itemNumber.take(20))
                ps.setString(2, source.description.take(40))
                ps.setString(3, source.locationCode.take(20))
                ps.setString(4, actualQty.toString().take(10))
                ps.setString(5, source.itemType.take(20))
                ps.setString(6, source.batchNumber.take(20))
                ps.setString(7, user.take(20))
                ps.setString(8, "Uml. aus -$actualQty".take(20))
                ps.executeUpdate()
            }

            conn.prepareStatement(
                """INSERT INTO dbo.journal
                   (Artikelnummer, Bezeichnung, Lagerort, Menge, Datum, Art, Lieferschein, Zusatz1, Bemerkung)
                   VALUES (?, ?, ?, ?, CAST(GETDATE() AS DATE), ?, ?, ?, ?)"""
            ).use { ps ->
                ps.setString(1, source.itemNumber.take(20))
                ps.setString(2, source.description.take(40))
                ps.setString(3, destLocationCode.take(20))
                ps.setString(4, actualQty.toString().take(10))
                ps.setString(5, source.itemType.take(20))
                ps.setString(6, source.batchNumber.take(20))
                ps.setString(7, user.take(20))
                ps.setString(8, "Uml. ein +$actualQty".take(20))
                ps.executeUpdate()
            }

            bookingId
        }
    }

    // ── Relocation revert ──

    fun getRecentRelocations(): List<RecentRelocation> {
        return SqlClient.withWms { conn ->
            conn.prepareStatement(
                """SELECT j_in.Artikelnummer, j_in.Bezeichnung, j_in.Lagerort AS dest,
                          j_out.Lagerort AS src, j_in.Menge, j_in.Zusatz1, j_in.Datum,
                          r.Rownumber AS destRowNum
                   FROM dbo.journal j_in
                   INNER JOIN dbo.journal j_out
                       ON j_out.Artikelnummer = j_in.Artikelnummer
                       AND j_out.Menge = j_in.Menge
                       AND j_out.Datum = j_in.Datum
                       AND j_out.Zusatz1 = j_in.Zusatz1
                       AND RTRIM(j_out.Bemerkung) LIKE 'Uml. aus%'
                       AND RTRIM(j_in.Bemerkung) LIKE 'Uml. ein%'
                   INNER JOIN dbo.rohteillager r
                       ON RTRIM(r.Artikelnummer) = RTRIM(j_in.Artikelnummer)
                       AND RTRIM(r.Lagerort) = RTRIM(j_in.Lagerort)
                       AND CAST(RTRIM(r.Menge) AS INT) > 0
                   WHERE RTRIM(j_in.Bemerkung) LIKE 'Uml. ein%'
                   ORDER BY j_in.Datum DESC"""
            ).use { ps ->
                ps.executeQuery().use { rs ->
                    val results = mutableListOf<RecentRelocation>()
                    val seen = mutableSetOf<Int>()
                    while (rs.next() && results.size < 20) {
                        val destRow = rs.getInt("destRowNum")
                        if (seen.add(destRow)) {
                            results.add(
                                RecentRelocation(
                                    destBookingId = destRow.toString(),
                                    itemNumber = rs.getString("Artikelnummer")?.trim() ?: "",
                                    description = rs.getString("Bezeichnung")?.trim() ?: "",
                                    fromLocation = rs.getString("src")?.trim() ?: "",
                                    toLocation = rs.getString("dest")?.trim() ?: "",
                                    qty = (rs.getString("Menge")?.trim() ?: "0").toIntOrNull() ?: 0,
                                    movedBy = rs.getString("Zusatz1")?.trim() ?: "",
                                    movedAt = rs.getTimestamp("Datum")?.toString() ?: ""
                                )
                            )
                        }
                    }
                    results
                }
            }
        }
    }

    fun revertRelocation(destBookingId: Int, originalSourceLocation: String, qty: Int, user: String): String {
        return SqlClient.withWms { conn ->
            val dest = conn.prepareStatement(
                "SELECT * FROM dbo.rohteillager WHERE Rownumber = ?"
            ).use { ps ->
                ps.setInt(1, destBookingId)
                ps.executeQuery().use { rs ->
                    if (rs.next()) rowToBooking(rs) else throw IllegalStateException("Destination booking not found")
                }
            }

            val actualQty = minOf(qty, dest.qty)
            val newDestQty = dest.qty - actualQty

            if (newDestQty <= 0) {
                conn.prepareStatement("DELETE FROM dbo.rohteillager WHERE Rownumber = ?").use { ps ->
                    ps.setInt(1, destBookingId)
                    ps.executeUpdate()
                }
            } else {
                conn.prepareStatement("UPDATE dbo.rohteillager SET Menge = ? WHERE Rownumber = ?").use { ps ->
                    ps.setString(1, newDestQty.toString())
                    ps.setInt(2, destBookingId)
                    ps.executeUpdate()
                }
            }

            val bookingId: String
            conn.prepareStatement(
                """INSERT INTO dbo.rohteillager
                   (Artikelnummer, Bezeichnung, Lagerort, Menge, Datum, Art, Lieferschein, Zusatz1)
                   VALUES (?, ?, ?, ?, CAST(GETDATE() AS DATE), ?, ?, ?)""",
                Statement.RETURN_GENERATED_KEYS
            ).use { ps ->
                ps.setString(1, dest.itemNumber.take(20))
                ps.setString(2, dest.description.take(40))
                ps.setString(3, originalSourceLocation.take(20))
                ps.setString(4, actualQty.toString().take(10))
                ps.setString(5, dest.itemType.take(20))
                ps.setString(6, dest.batchNumber.take(20))
                ps.setString(7, user.take(20))
                ps.executeUpdate()
                bookingId = ps.generatedKeys.use { keys ->
                    if (keys.next()) keys.getInt(1).toString() else "0"
                }
            }

            conn.prepareStatement(
                """INSERT INTO dbo.journal
                   (Artikelnummer, Bezeichnung, Lagerort, Menge, Datum, Art, Lieferschein, Zusatz1, Bemerkung)
                   VALUES (?, ?, ?, ?, CAST(GETDATE() AS DATE), ?, ?, ?, ?)"""
            ).use { ps ->
                ps.setString(1, dest.itemNumber.take(20))
                ps.setString(2, dest.description.take(40))
                ps.setString(3, dest.locationCode.take(20))
                ps.setString(4, actualQty.toString().take(10))
                ps.setString(5, dest.itemType.take(20))
                ps.setString(6, dest.batchNumber.take(20))
                ps.setString(7, user.take(20))
                ps.setString(8, "Storno Uml. -$actualQty".take(20))
                ps.executeUpdate()
            }

            conn.prepareStatement(
                """INSERT INTO dbo.journal
                   (Artikelnummer, Bezeichnung, Lagerort, Menge, Datum, Art, Lieferschein, Zusatz1, Bemerkung)
                   VALUES (?, ?, ?, ?, CAST(GETDATE() AS DATE), ?, ?, ?, ?)"""
            ).use { ps ->
                ps.setString(1, dest.itemNumber.take(20))
                ps.setString(2, dest.description.take(40))
                ps.setString(3, originalSourceLocation.take(20))
                ps.setString(4, actualQty.toString().take(10))
                ps.setString(5, dest.itemType.take(20))
                ps.setString(6, dest.batchNumber.take(20))
                ps.setString(7, user.take(20))
                ps.setString(8, "Storno Uml. +$actualQty".take(20))
                ps.executeUpdate()
            }

            bookingId
        }
    }

    // ── FIFO lookup ──

    fun searchFifo(itemNumber: String): List<Booking> {
        return SqlClient.withWms { conn ->
            conn.prepareStatement(
                """SELECT * FROM dbo.rohteillager
                   WHERE RTRIM(Artikelnummer) = ? AND CAST(RTRIM(Menge) AS INT) > 0
                   ORDER BY Datum ASC, Rownumber ASC"""
            ).use { ps ->
                ps.setString(1, itemNumber)
                ps.executeQuery().use { rs ->
                    val results = mutableListOf<Booking>()
                    while (rs.next()) results.add(rowToBooking(rs))
                    results
                }
            }
        }
    }

    // ── Zone / shelf queries ──

    fun getZoneOccupancy(shelfCodes: List<String>): Int {
        if (shelfCodes.isEmpty()) return 0
        val placeholders = shelfCodes.joinToString(" OR ") { "RTRIM(Lagerort) LIKE ?" }
        return SqlClient.withWms { conn ->
            conn.prepareStatement(
                """SELECT COUNT(DISTINCT RTRIM(Lagerort))
                   FROM dbo.rohteillager
                   WHERE CAST(RTRIM(Menge) AS INT) > 0 AND ($placeholders)"""
            ).use { ps ->
                shelfCodes.forEachIndexed { i, sc -> ps.setString(i + 1, "$sc-%") }
                ps.executeQuery().use { rs ->
                    if (rs.next()) rs.getInt(1) else 0
                }
            }
        }
    }

    fun getOccupiedCells(shelfCode: String): List<String> {
        return SqlClient.withWms { conn ->
            conn.prepareStatement(
                """SELECT DISTINCT RTRIM(Lagerort)
                   FROM dbo.rohteillager
                   WHERE RTRIM(Lagerort) LIKE ? AND CAST(RTRIM(Menge) AS INT) > 0"""
            ).use { ps ->
                ps.setString(1, "$shelfCode-%")
                ps.executeQuery().use { rs ->
                    val cells = mutableListOf<String>()
                    while (rs.next()) cells.add(rs.getString(1).trim())
                    cells
                }
            }
        }
    }

    fun getCellSlotCounts(shelfCode: String): Map<String, Int> {
        return SqlClient.withWms { conn ->
            conn.prepareStatement(
                """SELECT RTRIM(Lagerort) AS loc, COUNT(*) AS cnt
                   FROM dbo.rohteillager
                   WHERE RTRIM(Lagerort) LIKE ? AND CAST(RTRIM(Menge) AS INT) > 0
                   GROUP BY RTRIM(Lagerort)"""
            ).use { ps ->
                ps.setString(1, "$shelfCode-%")
                ps.executeQuery().use { rs ->
                    val map = mutableMapOf<String, Int>()
                    while (rs.next()) map[rs.getString("loc").trim()] = rs.getInt("cnt")
                    map
                }
            }
        }
    }

    // ── Helpers ──

    private fun rowToBooking(rs: ResultSet): Booking {
        return Booking(
            id = rs.getInt("Rownumber").toString(),
            locationCode = rs.getString("Lagerort")?.trim() ?: "",
            itemNumber = rs.getString("Artikelnummer")?.trim() ?: "",
            itemType = rs.getString("Art")?.trim() ?: "",
            batchNumber = rs.getString("Lieferschein")?.trim() ?: "",
            qty = (rs.getString("Menge")?.trim() ?: "0").toIntOrNull() ?: 0,
            description = rs.getString("Bezeichnung")?.trim() ?: "",
            bookedBy = rs.getString("Zusatz1")?.trim() ?: "",
            bookedAt = rs.getTimestamp("Datum")?.toString() ?: "",
            status = "Active"
        )
    }
}
